---
title: "Analysis of BLAST Results in Distinct Age Groups and Sample Types"
author: "Leila Esmaeili"
date: "October 23, 2018"
output: github_document
Contact: lesmaeili@dons.usfca.edu
---

# Introduction
With a new  understanding that bacterial communites on human skin are
personalized, scientists have been curious to find out if this new discovery can 
be manipulated for forensic applications. For example, we want to know if we can
identify a criminal matching his/her skin bacterial colonies with that of those
objects he/she may have touched. With the computations that I have done 
with the data from the assigned artilce, Fierir et al., I
have explored the questions if a person's age can be determined using the 
bacteria on the skin.  Also, I wanted to understand the results from  dust vs
sebum samples.

# Methods
First, The authors compared bacteria on individual keys of three computer 
keyboards to those found on the fingers of the keyboard owners. Second, we
examined the similarity between skin-associated bacterial communities on objects
stored at −20 °C(a standard method for storing samples before DNA extraction) 
versus those objects stored under typical indoor environmental con-ditions for
up to 14 days. Finally, we linked objects to specific individuals by comparing 
the bacteria on their computer mice against a database containing bacterial
community information for more than 250 hand surfaces, including the hand of
the owner.

Their method of doing this was very simple. They just swabbed bacteria on the
surface of keyboards and computer mice and the fingers of people touching them.  
They then grew these bacteria in petri dishes and isolated communites.  And,
then they extracted DNA and sequences the DNA and finall compared their 
sequences to a natiaonl data base.

## Sample origin and sequencing
DNA extraction was done using the MO BIO PowerSoil DNA Isolation kit.  
PCR was used for DNA amplification.  Pysrosequencing was carried out on 454 
Life Sciences Genome Sequencer FLX (Roche) at University of South Carolina.

## Computational
The process we used to analyze data is the following:

Also, the process that we are sending the data through is:
1. Download the Data
2. Apply fastQC to the data which stands for fast Quality Check
3. Trim the sequences
4. Convert from fastq to fasta format
5. Blastn to look at which bacteria was most similar

In our analysis, we used only one part of the data set relevant to us (i.e. mice
and fingertip dataset).  With our R studio codes, we have created a large table where all the information from above analysis are tabulated. Our codes allow for creating
data, generating reports and creating plots to answer our questions. Let's go 
over a specific code as an example.
+ #joined_blast_data_metadata %>%
+ # filter(age_s == 32) %>%  # filtering based on specfic column
+ #  ggplot(aes(x = pident)) +  # this is for x axis
+ #    geom_histogram() +
+ #   ggtitle("Percent Identiy Match age32") + # This is our title
+ #    xlab("Percent") # this is our x axis label

In the above code, the filter command was used to choose 32 year olds in the 
age coloumn and draw a histogram with percent identity on  the x axis and xlab
command is used for to label.  I used the same code to figure out the difference
between dust and sebum. 

# Results
The results of the histograms indicate there are distinct differences between 
bacterial colonies for each age group (Figure 1, 2). In addition, the 
differnces between dust and sebum (Figure 3, 4) have been very also graphed in 
a histogram.  In Figure 2,  there is a strange peak at around 87 
percent which is unsual. We see similar unusual peak in figures 3 and 4. 
```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the install.packages() function
# or with the 'Packages' pane in RStudio

# load packages
library("dplyr")  # This is for tables
library("tidyr") # creates tidy data
library("knitr") # generates reports
library("ggplot2") # creates plot
```

```{r make-read-in-data-function}
# Output format from BLAST is as detailed on:
# https://www.ncbi.nlm.nih.gov/books/NBK279675/
# In this case, we used: '10 sscinames std'
# 10 means csv format
# sscinames means unique Subject Scientific Name(s), separated by a ';'
# std means the standard set of result columns, which are:
# 'qseqid sseqid pident length mismatch
# gapopen qstart qend sstart send evalue bitscore',


# this function takes as input a quoted path to a BLAST result file
# and produces as output a dataframe with proper column headers
# and the 'qseqid' column split into sample and seq number
read_blast_output <- function(filename) {
  data_in <- read.csv(filename,
                      header = FALSE, # files don't have column names in them
                      col.names = c("sscinames", # unique Subject Sci Name(s)
                                    "qseqid",    # Query Seq-id
                                    "sseqid",    # Subject Seq-id
                                    "pident",    # Percntge of identical matches
                                    "length",    # Alignment length
                                    "mismatch",  # Number of mismatches
                                    "gapopen",   # Number of gap openings
                                    "qstart",    # Start of alignment in query
                                    "qend",      # End of alignment in query
                                    "sstart",    # Start of alignment in subj
                                    "send",      # End of alignment in subject
                                    "evalue",    # Expect value
                                    "bitscore"))  # Bit score

  # Next we want to split the query sequence ID into
  # Sample and Number components so we can group by sample
  # They originally look like "ERR1942280.1"
  # and we want to split that into two columns: "ERR1942280" and "1"
  # we can use the separate() function from the tidyr library to do this
  # Note that we have to double escape the period for this to work
  # the syntax is
  # separate(column_to_separate,
  # c("New_column_name_1", "New_column_name_2"),
  # "seperator")
  data_in <- data_in %>%
    separate(qseqid, c("sample_name", "sample_number"), "\\.")
}
```

```{r read-in-BLAST-data}
# this makes a vector of all the BLAST output file names, including
# the name(s) of the directories they are in
files_to_read_in <- list.files(path = "output/blast",
                               full.names = TRUE)

# We need to create an empty matrix with the right number of columns
# so that we can rbind() each dataset on to it
joined_blast_data <- matrix(nrow = 0,
                            ncol = 14)

# now we loop over each of the files in the list and append them
# to the bottom of the 'joined_blast_data' object
# we do this with the rbind() function and the function we
# made earlier to read in the files, read_blast_output()
for (filename in files_to_read_in) {
  joined_blast_data <- rbind(joined_blast_data,
                             read_blast_output(filename))
}
```

```{r read-in-metadata-and-join}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE)

# Finally we use the left_join() function from dplyr to merge or 'join' the
# combined data and metadata into one big table, so it's easier to work with
# in R the `by = c("Run_s" = "sample_name")` syntax tells R which columns
# to match up when joining the datasets together
joined_blast_data_metadata <- metadata_in %>%
  left_join(joined_blast_data,
            by = c("Run_s" = "sample_name"))
```


```{r create-gender-column}
joined_blast_data_metadata <- joined_blast_data_metadata %>%
  mutate(host_subject_id_sex = substr(host_subject_id_s, 1, 1))
```


```{r unique-values-in-column}
unique(as.character(joined_blast_data_metadata$host_subject_id_s))
```


```{r histograms_percent_identity_match _age32}
# Here we're using the dplyr piping syntax to select a subset of rows matching a
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram.

joined_blast_data_metadata %>%
  filter(age_s == 32) %>%  # filtering based on specfic column age
    ggplot(aes(x = pident)) +  # this is for x axis
    geom_histogram() +
    ggtitle("Percent Identiy Match age 32-Figure 1") + # This is our title
    xlab("Percent") # this is our x axis label
```
```{r histograms_percent_identity_match_age25}
# Here we're using the dplyr piping syntax to select a subset of rows matching
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram.
joined_blast_data_metadata %>%
  filter(age_s == 25) %>%
  ggplot(aes(x = pident)) +
    geom_histogram() +
    ggtitle("Percent Identity Match age 25:Figure 2") +
    xlab("Percent")

``` 

```{r Processing for Graph2}
colnames(joined_blast_data_metadata)
joined_blast_data_metadata["env_material_s"]
unique(joined_blast_data_metadata["env_material_s"])
```

```{r histograms_dust}
# Here we're using the dplyr piping syntax to select a subset of rows matching
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram.
joined_blast_data_metadata %>%
filter(env_material_s == "dust") %>%
  ggplot(aes(x = pident)) +
  geom_histogram() +
  ggtitle("Percent Identity Match dust:Figure3") +
  xlab("Percent")
```

```{r histograms_sebum}
# Here we're using the dplyr piping syntax to select a subset of rows matching
# criteria we specify (using the filter) function, and then pull out a column
# from the data to make a histogram.
joined_blast_data_metadata %>%
filter(env_material_s == "sebum") %>% # filtering for sebum
  ggplot(aes(x = pident)) + # This is x-axis
  geom_histogram() +
  ggtitle("Percent Identity Match sebum:Figure 4") +
  xlab("Percent")
```


# Discussion
After I did the command to identify unique age groups, I then plotted for two of
the unique age groups a histogram with counts on the Y-axis and percent identity 
match on the X-axis. The age histograms allow us to compare percent identity
match between age 32 and  25.  We can see that histograms between 
these age groups can easily be  distinguished. In other words, looking a given
histogram, we can easily conclude if the person who has touched the mouse is
32 or  25. However, to make a conclusion that each age group gravitates towards
certain colonies is premature. We need more data to be able to make that 
conclustion. Only 60 counts showed 100% identity match for age 32 whereas 
100 counts showed 100% identity match for age 25.  The unusal peak at 85 is 
probably due to presence of mite. It is important to remember that identity
match lower than 97% percent doesnt really count as a match. It is probably 
an error.

In addiiton, I plotted histograms of counts vs percent indentity match for Dust
vs Sebum (figures 3 and 4). Dust refers to bacterial colonies on mice while 
sebum refers to bacterial colonies on fingers.  Over 2000 counts showed 100% 
identity match for sebum where as only about 1500 counts showed 100% mathces for
dust. The unusal peak at 85 could be due to presence of mite on the mouse.

# Conclusion 
Looking at the Figures 1, 2 (age) and Figures 3, 4 (environmental samples ) has 
given helpful information on understanding data produced by blast. Further 
studies are needed to determine if these findings can be applied in forensic
and courts for applications in legal systems.

